src/tooling/tooling-engine-singleton.md

```diff
@@ -0,0 +1,123 @@
+  ~ License, v. 2.0. If a copy of the MPL was not distributed with this
+  ~ file, You can obtain one at https://mozilla.org/MPL/2.0/.
+-->
+
+# Tooling with Engine Singletons
+
+It is important for you to understand the [Singleton (anti)-pattern](https://en.wikipedia.org/wiki/Singleton_pattern) to properly utilize this system.
+
+An Engine Singleton is registered through a specific Godot-provided Singleton.
+
+Engine Singletons in Godot:
+
+- are Object types
+- are always accessible to GDScript (and other bound langs like C#)
+- must be manually registered and unregistered in the `LevelInit::Scene` step
+
+
+## Defining a Singleton
+
+Defining a singleton is the same as the Simple Type for an Object Type.
+
+```rust
+#[derive(GodotClass)]
+#[class(tool, base=Object)]
+struct MyEditorSingleton {
+    #[base]
+    base : Base<Object>
+}
+
+#[godot_api]
+impl IObject for MyEditorSingleton {
+    fn init(base : Base<Object>) -> Self {
+
+        // you may want to omit this by adding "init" to the #[class(...)]
+        Self {
+            base
+        }
+    }
+}
+
+#[godot_api]
+impl MyEditorSingleton {
+
+    #[func]
+    fn foo(&mut self) {}
+
+}
+```
+
+
+## Registering a singleton
+
+Registering singletons is done during the `InitLevel::Scene` stage of initialization.
+
+
+```rust
+struct MyExtension;
+
+unsafe impl ExtensionLibrary for MyExtension {
+
+    // these functions are optional if you are not using an 
+    /// Engine Singleton or similarly manually registred types
+    fn on_level_init(level: InitLevel) {
+        if level == InitLevel::Scene {
+
+            // the string name is how your singleton is accessed
+            // make sure it is easy to understand!
+            Engine::singleton().register_singleton(
+                StringName::from("MyEditorSingleton"),
+                MyEditorSingleton::alloc_gd().upcast(),
+            );
+        }
+    }
+
+    fn on_level_deinit(level: InitLevel) {
+        if level == InitLevel::Scene {
+
+            // make sure to unregister as well or you'll get a mess of warnings
+            // especially if you enable hot reloading
+            Engine::singleton().unregister_singleton(
+                StringName::from("MyEditorSingleton")
+            );
+        }
+    }
+}
+```
+
+
+## Calling from GDScript
+
+Now that your Engine Singleton is available (and once you've rebuilt and reloaded!) you should be able to access the singleton like so
+
+```Python
+extends Node
+
+func _ready() -> void:
+    MyEditorSingleton.foo()
+
+```
+
+
+## Accessing the SceneTree from your Singleton
+
+Because the singleton is an Object type, it is not connected to the SceneTree in the same way a Node type would be. However it is still possible to
+get access to the scene tree if you so desire. Here's a snippet of how to do this, which uses an external public function so it's easily accessible
+in case you need multiple calls or even multiple singletons!
```

Comment:
This wording is confusing: being a `Node` (or rather, inheriting it) doesn't automatically give access to the scene tree. The object would still need to be added to the tree.

I would just omit the first sentence and instead directly focus on getting scene tree access.

Reply:
Maybe more generally, I'm not sure if this is a workflow we should promote? The scene will not be loaded at the time the singleton is instantiated, so what can one meaningfully do with it?

Typically, it's better to add special nodes as parts of scenes.

Reply:
> I'm not sure if this is a workflow we should promote?

That's true. I think it's a fairly hacky way to do things. Maybe a better approach would be to promote using `EditorPlugin` to add an autoload singleton Node? At least that's the intended way to have a singleton contained in the SceneTree for the duration of runtime

Reply:
That's a good idea -- maybe autoloads could even be a separate recipe. But one thing at a time 🙂